Add support for IN2 rule

alecalvop opened the pull request.

alecalvop committed:
Add support for IN2 rule

alecalvop commented on src/analysis/parsing/parser.rs:
```diff
@@ -369,8 +369,14 @@ impl <'a> FileParser<'a> {
                 TokenKind::Whitespace => {
-                    // TODO: this might not work correctly with tabs
-                    self.current_column += self.lexer.slice().len() as u32;
+                    let slice = self.lexer.slice();
+                    for ch in slice.chars() {
+                        if ch == '\t' {
+                            self.current_column += 4;
+                        } else {
+                            self.current_column += 1;
+                        }
+                    }
```
For this rule, we encountered a problem where the tabs were assigned a column width value of 1. This made the indentation checks on rule in3 break, because it would think the line is indented lesser that it should have.

I drafted this quick fix to showcase the issue, but another solution could be to create a new TokenKind for Tabs and separate it to the whitespaces also.

Another thing is that I just put a magic number of 4 as the value, but I'm not sure if this would be configurable or this should be tied to the linting parameter for indentation.

I would like to get your input about this @JonatanWaern

JonatanWaern replied:
I am fairly certain that the code was already correct. Modifying this will most likely break vscode decoding of file positions as reported by the DLS. The tab character is after all just one character, even if on display it appears wider.

I think I sort-of disagree with the separation of tabs from whitespace in general, too. Instead I think the correct solution is to, at the point where you want to know your 'real' identation, read the file contents of the prefixrange of the first token on your line and calculate the displayed indentation off of that.

The indentation should probably be configurable, exactly where that option should be I am less sure of. At some point the server should supply things like auto-indentation of lines and regions of code, so it kinda depends on whether that goes under the 'linting' umbrella or not. I suspect it doesnt, so I would add that option as a top-level server configurable with a default of 4.

alecalvop replied:
I also thought that vscode would count the TAB character as just one character, but it seems that it counts it as 4, which actually currently breaks the positions reported by the DLS when the file is using TABs for indentation.
![image](https://github.com/user-attachments/assets/c2248630-9280-4fb6-adfc-ce3dde79644f)
The example above shows that currently when using TABs the DLS is reporting different locations compared to the vscode positioning. The most clear example is the long line, where the yellow warning mark is further right than it should be, it should be right next to the ruler that indicates the 80 char mark.
It seems that this also impacts syntax analysis, as seen on the error reported on line 11: it is reporting Col 32, but the error is displayed in Col 38.

JonatanWaern replied:
I have dug into this further and am getting inconsistent results compared to you.

For, for example syntax errors, I am getting the correct underlines (here on the token following a missing comma).
Curiously, attempting the above fix (2nd screenshot) breaks token-reading more generally (the server things there is an error at the log-kind, because it incorrectly reads later than it should). In your case isnt the report at col 32 correct, and it being shown at col 38 just a result of the tabs being wider on display?

![image](https://github.com/user-attachments/assets/8b24c57d-06e3-40cb-9b65-3673e6922db5)
![image](https://github.com/user-attachments/assets/27ecac05-d44d-4062-adc9-e636bd38f803)

Rather, I think the error is that when the long-lines rule checks for line-length it just checks the length of the str representing the line, not considering the tab characters as longer when displayed. And since this rule actually does deal with display, the result becomes incorrect. My suggestion is to add a linter option for how long it should consider tabs to be (default 4) and use that to calculate the "real" line lengths for that rule in particular (though some care has to be taken when reporting the lint warning, as that should be done using regular file positioning)

EDIT:
What I observe in the same syntax highlighting scenario, I dont think there is an offset issue here.
![image](https://github.com/user-attachments/assets/f66b9f03-b994-4906-8b49-21f8ca7e1534)

alecalvop committed:
Add post process method for linting error

This post process method removes other errors if
they are in the same line ad in2 errors.

This is because in2 is reported when there are tabs in the
document, which causes miss calculations in indentation
levels.

alecalvop commented on src/lint/mod.rs:
```diff
@@ -126,16 +130,34 @@ pub fn begin_style_check(ast: TopAst, file: String, rules: &CurrentRules) -> Res
+        !in2_ranges.iter().any(|range|
+            (range.row_start == error.range.row_start || range.row_end == error.range.row_end)
+            && error.description != IN2Rule::description())
+    });
```
It would be better to have an error type to remove only indentation rules when in2 is hit

jvsqzj committed:
Merge branch 'feature/55-indentation-rules' into feat/in2

jvsqzj committed:
Wrap LocalDMLError in DMLStyleError to hold local Lint module info

jvsqzj committed:
Use enum RuleType to as identifier for Rules instead of name()

jvsqzj committed:
Merge pull request #7 from kevv87/dml-style-err-types

Wrap LocalDMLError in DMLStyleError to hold local Lint module info

alecalvop committed:
Remove unused imports

JonatanWaern commented on src/lint/rules/indentation.rs:
```diff
@@ -56,16 +56,19 @@ impl LongLinesRule {
+                    range: Range::<ZeroIndexed>::from_u32(rowu32, rowu32, self.max_length, len),
+                    description: msg,
+                },
+                rule_type: Self::get_rule_type(),
```
You could automate the wrapping of LocalDMLError -> DMLStyleError by letting evaluate rules call into a trait-provided function that calls check() and then wraps the results with the values from get_rule_type(). (the benefit of applying this change is ATM minimal, but it does more strongly connect the get_rule_type semantic into the functionality)

alecalvop replied:
Yes, actually Kevin is working on adding the creation of errors into the rule trait.

JonatanWaern commented on src/lint/mod.rs:
```diff
@@ -124,20 +132,38 @@ impl LinterAnalysis {
+        .map(|style_err| style_err.error.range)
+        .collect();
+
+    // Remove linting errors that are in in2 rows
```
I dont know the specifics of how in2 causes conflicts with other lints, but if this is expected to be a more general problem it could be an idea to let each rule have a "priority", and report only the highest priority rule in each range.

You could also consider using some sort of span-tree structure to perform this logic on insert rather than double-iteration done here, but I would leave such optimizations until you notice the linting actually taking significant time on large devices (likely the time will always be dominated by the parser anyway)
